Support Non-Boolean and Mixed Contraints

s9latimm opened the pull request:
- [x] Implement constraint types (boolean, non-boolean, mixed)
- [x] Update parsing / builder / transaction
- [x] Update FM-Writer
- [x] Add new unit / integration tests for constraints
- [x] Pass existing tests / CI

s9latimm committed:
Support non-boolean and mixed contraints

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	lib/Feature/XmlConstants.h
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/UnittestHelper.h

s9latimm committed:
Clean up template

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Merge branch 'vara-dev' into f-MixedConstraints

s9latimm committed:
Add tests

M	include/vara/Feature/FeatureModelWriter.h
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/resources/CMakeLists.txt
M	unittests/resources/xml/test_children.xml
A	unittests/resources/xml/test_constraints.xml

s9latimm committed:
Fix python tests

M	bindings/python/tests/TEST_INPUTS/example_feature_model.xml
M	bindings/python/tests/TEST_INPUTS/simple_example_feature_model.xml
M	bindings/python/tests/TEST_INPUTS/simple_example_feature_model_merge.xml

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -34,8 +34,6 @@ class FeatureModel {
   using FeatureMapTy = llvm::StringMap<std::unique_ptr<Feature>>;
-  using ConstraintTy = Constraint;
-  using ConstraintContainerTy = std::vector<std::unique_ptr<ConstraintTy>>;
   using RelationshipContainerTy = std::vector<std::unique_ptr<Relationship>>;
```
Ok, maybe I don't get the design, but why is constraint gone? But a class added below?

vulder reviewed (changes requested):
Could you please add a bit of documentation so it's clearer what what does and how stuff works.

s9latimm committed:
Refactor

M	include/vara/Feature/FeatureModel.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Remove TODO

M	lib/Feature/FeatureModelWriter.cpp

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-MixedConstraints

M	bindings/python/vara-feature/pybind_FeatureModelXmlWriter.cpp
M	bindings/python/vara-feature/pybind_FeatureSourceRange.cpp
M	include/vara/Configuration/Configuration.h
M	include/vara/Feature/ConstraintParser.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelWriter.h
M	include/vara/Feature/FeatureSourceRange.h
M	include/vara/Feature/StepFunctionParser.h
M	lib/Configuration/Configuration.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelTransaction.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	lib/Sampling/SampleSetWriter.cpp
M	unittests/Configuration/Configuration.cpp
M	unittests/Configuration/ConfigurationOption.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/Feature/FeatureSourceRange.cpp
M	unittests/Sampling/SampleSetParserTests.cpp
M	unittests/Sampling/SampleSetWriterTests.cpp

s9latimm committed:
Fix enums

M	include/vara/Feature/FeatureModel.h
